gpm_core/src/lockfile.rs

```diff
@@ -0,0 +1,120 @@
+use anyhow::Context;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::path::{Path, PathBuf};
+
+/// Allow to look for a specific version of a dependancies
+#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
+#[serde(tag = "type")]
+pub enum LockSource {
+    /// Use the mod with the given ID and the matching version.
+    ///
+    /// Even if the parent contain the id, we still store it there, as it may be usefull that id
+    /// differ from it's parent.
+    IdVersion { identifier: String, version: String },
+    /// Use a specific path on the local filesystem. If relative, it'll be based around the
+    /// profile folder.
+    Path { path: PathBuf },
+}
+
+/// contain a fixed set of mod depency, with each depency having a specific version. Mod are
+/// identified by their id. They are unique.
+#[derive(Serialize, Deserialize, Default, Debug)]
+pub struct LockFile {
+    pub dependencies: HashMap<String, LockSource>,
+}
+
+impl LockFile {
+    /// create a new [`LockFile`] with no dependancies.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// return the [`LockSource`] corresponding to a given package identifier if it is present in
+    /// this [`LockFile`], None otherwise.
+    pub fn depency_source(&self, identifier: &str) -> Option<LockSource> {
+        self.dependencies.get(identifier).cloned()
+    }
+
+    /// define the [`LockSource`] for a given package identifier, overwriting the current one.
+    ///
+    /// return the previous [`LockSource`] if overwriting it
+    pub fn set_depency_source(
+        &mut self,
+        identifier: String,
+        source: LockSource,
+    ) -> Option<LockSource> {
+        self.dependencies.insert(identifier, source)
+    }
+
+    /// remove the mod with the given id. Return the previous entry if it exist, None otherwise.
+    pub fn remove_depency_source(&mut self, identifier: &str) -> Option<LockSource> {
+        self.dependencies.remove(identifier)
+    }
+
+    /// iterate over all the locked dependancies
+    pub fn iter_depency_source(&self) -> std::collections::hash_map::Iter<'_, String, LockSource> {
```

Comment:
are those function named `depency` on purpose or is it a typo because they act on the depen**den**cy list, missing `den` in `dependency`.

Reply:
indeed (this word seem particularly complex to me for some reason)